math/src/stableswap/math.rs

```diff
@@ -183,23 +270,101 @@ pub fn calculate_withdraw_one_asset<const N: u8, const N_Y: u8>(
-
 	let fee = dy_0.checked_sub(dy)?;
 
-	Some((dy, fee))
+	let amount_out = normalize_value(dy, TARGET_PRECISION, asset_out_decimals, Rounding::Down);
+	let fee = normalize_value(fee, TARGET_PRECISION, asset_out_decimals, Rounding::Down);
+	Some((amount_out, fee))
+}
+
+pub fn calculate_add_one_asset<const D: u8, const Y: u8>(
+	reserves: &[AssetReserve],
+	shares: Balance,
+	asset_index: usize,
+	share_asset_issuance: Balance,
+	amplification: Balance,
+	fee: Permill,
+) -> Option<Balance> {
+	if share_asset_issuance.is_zero() {
+		return None;
+	}
+	if asset_index >= reserves.len() {
+		return None;
+	}
+	if shares > share_asset_issuance {
+		return None;
+	}
+	let n_coins = reserves.len();
+	if n_coins <= 1 {
+		return None;
+	}
+
+	let asset_in_decimals = reserves[asset_index].decimals;
+	let reserves = normalize_reserves(reserves);
+
+	let initial_d = calculate_d_internal::<D>(&reserves, amplification)?;
+	let (shares_hp, issuance_hp, d_hp) = to_u256!(shares, share_asset_issuance, initial_d);
+
+	let d1 = d_hp.checked_add(shares_hp.checked_mul(d_hp)?.checked_div(issuance_hp)?)?;
+
+	let xp: Vec<Balance> = reserves
+		.iter()
+		.enumerate()
+		.filter(|(idx, _)| *idx != asset_index)
+		.map(|(_, v)| *v)
+		.collect();
+
+	let y = calculate_y_internal::<Y>(&xp, Balance::try_from(d1).ok()?, amplification)?;
+
+	let fee = FixedU128::from(fee);
+	let xp_hp: Vec<U256> = reserves.iter().map(|v| to_u256!(*v)).collect();
+	let y_hp = to_u256!(y);
+
+	let mut reserves_reduced: Vec<Balance> = Vec::new();
+	let mut asset_reserve: Balance = Balance::zero();
+
+	for (idx, reserve) in xp_hp.iter().enumerate() {
+		let dx_expected = if idx == asset_index {
+			y_hp.checked_sub(reserve.checked_mul(d1)?.checked_div(d_hp)?)?
+		} else {
+			reserve.checked_mul(d1)?.checked_div(d_hp)?.checked_sub(*reserve)?
+		};
+
+		let expected = Balance::try_from(dx_expected).ok()?;
+		let reduced = Balance::try_from(*reserve)
+			.ok()?
+			.checked_sub(fee.checked_mul_int(expected)?)?;
+
+		if idx != asset_index {
+			reserves_reduced.push(reduced);
+		} else {
+			asset_reserve = reduced;
+		}
+	}
+	let y1 = calculate_y_internal::<Y>(&reserves_reduced, Balance::try_from(d1).ok()?, amplification)?;
+	let dy = y1.checked_sub(asset_reserve)?;
+	/*
+	let dy = asset_reserve.checked_sub(y1)?;
+	let dy_0 = reserves[asset_index].checked_sub(y)?;
+	let fee = dy_0.checked_sub(dy)?;
+	 */
+	let amount_in = normalize_value(dy, TARGET_PRECISION, asset_in_decimals, Rounding::Down);
+	Some(amount_in)
+}
+pub fn calculate_d<const D: u8>(reserves: &[AssetReserve], amplification: Balance) -> Option<Balance> {
+	let balances = normalize_reserves(reserves);
+	calculate_d_internal::<D>(&balances, amplification)
 }
 
 /// amplification * n^n where n is number of assets in pool.
 pub(crate) fn calculate_ann(len: usize, amplification: Balance) -> Option<Balance> {
-	(0..len).try_fold(amplification, |acc, _| acc.checked_mul(len as u128))
+	amplification.checked_mul(len as u128)
```

Comment:
comment doesn't fit anymore, change is intentional?

Reply:
ah yes. this has changed. i will update